src/face_op.cpp

```diff
@@ -271,33 +271,42 @@ void Manifold::Impl::FlattenFaces() {
            [](auto& v) { v.store(0); });
 
   for_each_n(policy, countAt(0_uz), numTri,
-             [&edgeFace, &vertDegree, this](size_t tri) {
+             [&edgeTri, &vertDegree, this](size_t tri) {
+               const auto& ref = meshRelation_.triRef[tri];
                for (const int i : {0, 1, 2}) {
                  const int edge = 3 * tri + i;
                  const int pair = halfedge_[edge].pairedHalfedge;
                  if (pair < 0) {
-                   edgeFace[edge] = kRemove;
+                   edgeTri[edge] = kRemove;
                    return;
                  }
-                 const auto& ref = meshRelation_.triRef[tri];
                  if (ref.SameFace(meshRelation_.triRef[pair / 3])) {
-                   edgeFace[edge] = kRemove;
+                   edgeTri[edge] = kRemove;
                  } else {
-                   edgeFace[edge] = (static_cast<uint64_t>(ref.meshID) << 32) +
-                                    static_cast<uint64_t>(ref.coplanarID);
+                   edgeTri[edge] = tri;
                    ++vertDegree[halfedge_[edge].startVert];
                  }
                }
              });
 
+  auto comp = [&edgeTri, this](size_t a, size_t b) {
+    if (edgeTri[a] == edgeTri[b]) return false;
+    if (edgeTri[a] == kRemove) return false;
+    if (edgeTri[b] == kRemove) return true;
+    const TriRef& refA = meshRelation_.triRef[edgeTri[a]];
+    const TriRef& refB = meshRelation_.triRef[edgeTri[b]];
+    return refA.meshID < refB.meshID || (refA.meshID == refB.meshID &&
+                                         (refA.coplanarID < refB.coplanarID ||
+                                          (refA.coplanarID == refB.coplanarID &&
+                                           refA.faceID < refB.faceID)));
+  };
+
   Vec<size_t> newHalf2Old(halfedge_.size());
   sequence(newHalf2Old.begin(), newHalf2Old.end());
-  stable_sort(
-      newHalf2Old.begin(), newHalf2Old.end(),
-      [&edgeFace](size_t a, size_t b) { return edgeFace[a] < edgeFace[b]; });
+  stable_sort(newHalf2Old.begin(), newHalf2Old.end(), comp);
   newHalf2Old.resize(std::find_if(countAt(0_uz), countAt(halfedge_.size()),
```

Comment:
btw I'm suddenly thinking if we should change this `find_if` into `lower_bound` to reduce time complexity.

Reply:
Good call - I updated a couple of similar functions I found in sort.cpp too.